### PR TITLE
Improve search

### PR DESCRIPTION
**Pull request by bennyz**
Now results are stored in a map which is then iterated and shown
in a satisfactory manner, example:
<pre>
repo: CQRSGolang
        https://api.github.com/repos/TechMaster/CQRSGolang
        stars: 0
repo: go-laboratory
        https://api.github.com/repos/liunian1004/go-laboratory
        stars: 0
</pre>

### CODE REVIEW

File: backend/rest/rest.go [hunk 1]
```diff
@@ -36,49 +36,61 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
...
 	opt := &github.SearchOptions{
 		ListOptions: github.ListOptions{PerPage: c.resultsPerPage},
 	}
 
-	var repos []string
 	results, _, err := c.Client.Search.Code(c.ctx, query, opt)
+	repos := make(map[int64]*backends.Result)
+
 	if err != nil {
 		log.Fatal("Failed to fetch results", err)
```
Comment (matanshukry) [inline backend/rest/rest.go]: break function flow after this line?

Comment (matanshukry) [inline backend/rest/rest.go:51]: break function flow after this line?

File: backend/rest/rest.go [hunk 2]
```diff
@@ -36,49 +36,61 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
...
 	}
 
-	for _, result := range results.CodeResults {
-		repos = append(repos, result.GetRepository().GetHTMLURL())
-		repoID := result.GetRepository().GetID()
-		fmt.Printf("Fetching value for: %v", repoID)
+	for _, codeResult := range results.CodeResults {
+		repoID := codeResult.GetRepository().GetID()
+		fmt.Printf("Fetching value for: %v\n", repoID)
```
Comment (matanshukry) [inline backend/rest/rest.go]: is there any log.debug or something like that?

File: backend/rest/rest.go [hunk 3]
```diff
@@ -36,49 +36,61 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
...
 		value, err := storage.Get(repoID)
 		fmt.Printf("Found value: %s for repoID: %v\n", value, repoID)
 
-		if err != nil || value == "" {
-			log.Printf("Value not found in DB, requesting amount of stars...\n")
+		if err != nil || value == nil {
 			repo, _, err := c.Client.Repositories.GetByID(c.ctx, repoID)
```
Comment (matanshukry) [inline backend/rest/rest.go]: Don't you want to log the error before overriding it?

File: backend/rest/rest.go [hunk 4]
```diff
@@ -36,49 +36,61 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
...
+
 	if err != nil {
 		log.Fatal("Failed to fetch results", err)
 	}
 
-	for _, result := range results.CodeResults {
-		repos = append(repos, result.GetRepository().GetHTMLURL())
-		repoID := result.GetRepository().GetID()
-		fmt.Printf("Fetching value for: %v", repoID)
+	for _, codeResult := range results.CodeResults {
```
Comment (matanshukry) [inline backend/rest/rest.go:55]: Can you do a multi-get for all the ids from storage?

**Review by matanshukry — CHANGES_REQUESTED** [review]

File: backend/rest/rest.go (hunk 1 above)
Comment (bennyz) [reply backend/rest/rest.go:51]: Fatal panics and exits

File: backend/rest/rest.go (hunk 2 above)
Comment (bennyz) [reply backend/rest/rest.go]: Nope, created issue #2 to address this

File: backend/rest/rest.go (hunk 3 above)
Comment (bennyz) [reply backend/rest/rest.go]: it's logged in db.go, but yeah, I should

File: backend/rest/rest.go (hunk 4 above)
Comment (bennyz) [reply backend/rest/rest.go:55]: what do you mean?

File: backend/rest/rest.go (hunk 1 above)
Comment (matanshukry) [reply backend/rest/rest.go:51]: exists app rather than function I'm assuming? is that the desired outcome?

File: backend/rest/rest.go (hunk 4 above)
Comment (matanshukry) [reply backend/rest/rest.go:55]: You're doing:
foreach (repo) { handleResult(storage.get(repo.id)); }

I suggest:
ids = repos.map(repo => repo.id).toList();
results = storage.get(ids);
handleResults(results);

File: backend/rest/rest.go (hunk 1 above)
Comment (bennyz) [reply backend/rest/rest.go:51]: no point keeping it running if we didn't get anything, no?

File: backend/rest/rest.go (hunk 4 above)
Comment (bennyz) [reply backend/rest/rest.go:55]: yeah, good idea

File: persistence/storage.go [hunk 5]
```diff
@@ -1,7 +1,12 @@
 package persistence
 
+// JSONValue holds a json representation of a struct
+type JSONValue []byte
+
 // Storage backend to be used by the persistence layer
 type Storage interface {
-	Save(key int64, value []byte) (int64, error)
-	Get(key int64) ([]byte, error)
+	Save(key int64, value JSONValue) (int64, error)
```
Comment (matanshukry) [inline persistence/storage.go:8]: What is the meaning of the int64 return value? document it

File: util/util.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package util
+
+import (
+	"strconv"
+	"strings"
+)
+
+// Difference returns the difference between two arrays
+func Difference(arr1, arr2 []int64) (result []int64) {
+	hash := make(map[int64]bool)
```
Comment (matanshukry) [inline util/util.go]: What happened to the empty struct idea? why not?

File: util/util_test.go [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package util
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestIntersection(t *testing.T) {
```
Comment (matanshukry) [inline util/util_test.go]: Rename to TestDifference

File: util/util.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
+package util
+
+import (
+	"strconv"
+	"strings"
+)
+
+// Difference returns the difference between two arrays
+func Difference(arr1, arr2 []int64) (result []int64) {
```
Comment (matanshukry) [inline util/util.go:9]: Why does the second argument is int64[] and the first has no declaration?

File: util/util.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	hash := make(map[int64]bool)
+
+	for _, item := range arr1 {
+		hash[item] = true
```
Comment (matanshukry) [inline util/util.go]: Hashing of an "object" - any considerations that should be taken care of?

File: persistence/sqlite/db.go [hunk 10]
```diff
@@ -53,18 +55,16 @@ func (s *sqlite) Save(key int64, value []byte) (int64, error) {
...
 	}
 
-	var value []byte
-	err = stmt.QueryRow(key).Scan(&value)
-	if err != nil {
-		log.Println(err)
-	} else {
-		log.Printf("Fetched for key %v", key)
+	var value []persistence.JSONValue
+	err = stmt.QueryRow(util.SliceToString(keys)).Scan(&value)
```
Comment (matanshukry) [inline persistence/sqlite/db.go:65]: Any chance we can avoid SliceToString()? That is, simply pass an array as the variable?

File: util/util.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+		if _, ok := hash[item]; !ok {
+			result = append(result, item)
+		}
+	}
+
+	return result
+}
+
+// SliceToString converts an int64 array to a comma separated string
+func SliceToString(arr []int64) string {
```
Comment (matanshukry) [inline util/util.go:26]: I'm not a fan of *Slice*ToString. It doesn't really slice anything, does it?

File: backend/rest/rest.go [hunk 12]
```diff
@@ -35,53 +38,56 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
 	return &client{githubClient, ctx, resultsPerPage}, nil
 }
 
-// Search searches code using the github api
-func (c *client) Search(query string, lang string) map[int64]*backends.Result {
+var repos map[int64]*backends.Result
```
Comment (matanshukry) [inline backend/rest/rest.go]: This is a file-scoped variable, yet it's between 2 functions. Is that standard way in GO?

File: backend/rest/rest.go [hunk 13]
```diff
@@ -35,53 +38,56 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
...
+	}
 
-		if err != nil || value == nil {
-			repo, _, err := c.Client.Repositories.GetByID(c.ctx, repoID)
+	values, err := storage.Get(repoIDs)
+	if err != nil {
+		fmt.Println(err)
+	}
 
+	cachedRepoIds := initializeCache(values)
```
Comment (matanshukry) [inline backend/rest/rest.go:64]: So the entire cache is to cache the calls to GitHub - what about memory cache to sqlite? should we even do it?

File: backend/rest/rest.go [hunk 14]
```diff
@@ -20,6 +22,7 @@ type client struct {
 }
 
 var storage persistence.Storage
+var cachedRepos []*backends.Result
```
Comment (matanshukry) [inline backend/rest/rest.go]: am I missing something or are you just populating this without actually using it?

**Review by matanshukry — CHANGES_REQUESTED** [review]

File: util/util.go (hunk 6 above)
Comment (bennyz) [reply util/util.go]: Will do

File: util/util.go (hunk 8 above)
Comment (bennyz) [reply util/util.go:9]: I learned a new trick :)
you can specify the type only once if all arguments are of the same type

File: persistence/sqlite/db.go (hunk 10 above)
Comment (bennyz) [reply persistence/sqlite/db.go:65]: I tried, it seems the driver cannot really handle arrays

File: util/util.go (hunk 11 above)
Comment (bennyz) [reply util/util.go:26]: Slice is the name of the data structure[1]

[1] - https://tour.golang.org/moretypes/7

File: backend/rest/rest.go (hunk 14 above)
Comment (bennyz) [reply backend/rest/rest.go]: leftover from a previous attempt, remove it

File: backend/rest/rest.go (hunk 13 above)
Comment (bennyz) [reply backend/rest/rest.go:64]: this is the memory cache, no?
I don't know if something more complex is required given this is a short-living program

File: backend/rest/rest.go (hunk 12 above)
Comment (bennyz) [reply backend/rest/rest.go]: done

File: persistence/storage.go (hunk 5 above)
Comment (bennyz) [reply persistence/storage.go:8]: I'll remove it

File: util/util.go (hunk 9 above)
Comment (bennyz) [reply util/util.go]: What do you mean?

File: util/util_test.go (hunk 7 above)
Comment (bennyz) [reply util/util_test.go]: Done

**Pull request merged**